services/yellow/src/api/Handlers/ChargeForOrderCancellationFeeHandler.cs

```diff
@@ -1,53 +1,62 @@
-        private readonly ICreditCardPaymentService _paymentService;
+        private readonly IPaymentOrderAdapter _paymentOrderAdapter;
 
         public ChargeForOrderCancellationFeeHandler(IOrderService orderService,
-            ICreditCardPaymentService paymentService)
+            IPaymentOrderAdapter paymentOrderAdapter)
         {
             _orderService = orderService;
-            _paymentService = paymentService;
+            _paymentOrderAdapter = paymentOrderAdapter;
         }
-
-        protected override async Task<OrderCancellationFeeCharged> Handle(ChargeForOrderCancellationFeeRequest message)
+        
+        public async Task Handle(ChargeForOrderCancellationFeeRequest message, IMessageHandlerContext context)
         {
-            Order order = await Ensure(message, m => _orderService.FindOneById(m.OrderId));
-
-            var amountToPay = order.Price.CancellationFee;
+            try
+            {
+                var validateResult = ValidateRequest(message);
+                if (!validateResult.IsValid())
+                    throw new ValidationException(validateResult);
+            
+                Order order = HandlerHelper.Ensure(await _orderService.FindOneById(message.OrderId))!;
+            
+                var amountToPay = order.Price.CancellationFee;
 
-            var chargeResult = (message.AlternativePaymentMethod is not null)
-                ? await ChargeOrder(order, amountToPay, message.AlternativePaymentMethod)
-                : await ChargeOrder(order, amountToPay);
+                var chargeResult = (message.AlternativePaymentMethod is not null)
+                    ? await _paymentOrderAdapter.ChargeOrder(order, amountToPay, message.AlternativePaymentMethod)
+                    : await _paymentOrderAdapter.ChargeOrder(order, amountToPay);
             
-            if (message.AlternativePaymentMethod is not null)
-                await AddPaymentMethodToOrder(order, message.AlternativePaymentMethod.CreditCard, chargeResult.AuthorizePaymentToken);
+                if (message.AlternativePaymentMethod is not null)
+                    await _paymentOrderAdapter.AddPaymentMethodToOrder(order, message.AlternativePaymentMethod.CreditCard, chargeResult.AuthorizePaymentToken);
 
-            // TODO : Should go in a separate handler (and should retrieve it from the used payment token)
-            var lastPaymentMethod = order.PaymentMethods!.Last();
-            await _orderService.AddTransaction(order, chargeResult.Transaction.GetOrderTransaction(lastPaymentMethod));
+                var lastPaymentMethod = order.PaymentMethods!.Last();
+                await _orderService.AddTransaction(order, chargeResult.Transaction.GetOrderTransaction(lastPaymentMethod));
 
-            return new OrderCancellationFeeCharged(message.OrderId);
+                await context.Publish(new OrderCancellationFeeCharged(message.OrderId));
+                await context.Reply(new Response<bool>(true));

```

Comment:
This `Response` possibly breaks the purple saga because it's not expecting a response.
`Response<T>` should be returned only to the ambassador's calls, as a quick-feedback.

Reply:
Suggestion applied